### PR TITLE
chore: Use Vite instead of Webpack for development

### PR DESCRIPTION
**Pull request by BrunoQuaresma**
Intentions:
- Make our dev process better
- Webpack is quite heavy and it was bugging my workspace
- Webpack is kinda complex to maintain
- Vite has better build time and DX

How to use:
```
CODER_HOST=https://dev.coder.com/ yarn vite
// or
CODER_HOST=https://dev.coder.com/ yarn dev:vite
```

I would appreciate it if people could test it and help me to catch any bugs before replacing webpack entirely.

### CODE REVIEW

**Review by kylecarbs — APPROVED** [review]
Assuming bundle-size is similar, I'm in support. Could we add data on how much faster it is?

File: site/index.html [hunk 1]
```diff
@@ -0,0 +1,37 @@
+<!DOCTYPE html>
```
Comment (jsjoeio) [inline site/index.html:1]: ❓ Why do we have to add this? (Or is this a file move? hard to tell)

File: site/package.json [hunk 2]
```diff
@@ -10,6 +10,7 @@
     "check:all": "yarn format:check && yarn lint && yarn test",
     "chromatic": "chromatic",
     "dev": "webpack-dev-server --config=webpack.dev.ts",
+    "dev:vite": "vite",
```
Comment (jsjoeio) [inline site/package.json:13]: I like that you're leaving webpack in and encouraging folks to try vite. Seems like the right move and then when we're confident Vite works, we can remove webpack 👍🏼

File: site/src/AppRouter.tsx [hunk 3]
```diff
@@ -12,21 +12,23 @@ import { XServiceContext } from "xServices/StateContext"
...
-import { AccountPage } from "./pages/UserSettingsPage/AccountPage/AccountPage"
-import { SecurityPage } from "./pages/UserSettingsPage/SecurityPage/SecurityPage"
-import { SSHKeysPage } from "./pages/UserSettingsPage/SSHKeysPage/SSHKeysPage"
-import { CreateUserPage } from "./pages/UsersPage/CreateUserPage/CreateUserPage"
-import { UsersPage } from "./pages/UsersPage/UsersPage"
-import { WorkspaceBuildPage } from "./pages/WorkspaceBuildPage/WorkspaceBuildPage"
-import { WorkspacePage } from "./pages/WorkspacePage/WorkspacePage"
-import { WorkspaceSchedulePage } from "./pages/WorkspaceSchedulePage/WorkspaceSchedulePage"
 
+const IndexPage = lazy(() => import("./pages"))
```
Comment (jsjoeio) [inline site/src/AppRouter.tsx:16]: Remind me again, why are we lazy loading? Speeds up things or something?

File: site/src/Main.tsx [hunk 4]
```diff
@@ -1,21 +1,7 @@
...
-if (process.env.NODE_ENV === "development" && process.env.INSPECT_XSTATE === "true") {
-  // configure the XState inspector to open in a new tab
-  inspect({
-    url: "https://stately.ai/viz?inspect",
-    iframe: false,
-  })
-  // configure all XServices to use the inspector
-  Interpreter.defaultOptions.devTools = true
-}
-
```
Comment (jsjoeio) [inline site/src/Main.tsx:18]: 👏🏼

File: site/tsconfig.json [hunk 5]
```diff
@@ -13,7 +13,7 @@
     "resolveJsonModule": true,
     "skipLibCheck": true,
     "strict": true,
-    "target": "es2018",
+    "target": "es2020",
```
Comment (jsjoeio) [inline site/tsconfig.json:16]: ❓reason for this? I like it but wondering if it was needed or happened in conjunction with Vite addition.

File: site/vite.config.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    port: process.env.PORT ? Number(process.env.PORT) : 8080,
+    proxy: {
+      "/api": {
+        target: process.env.CODER_HOST || "http://localhost:3000",
+        ws: true,
+        secure: process.env.NODE_ENV === "production",
+      },
+    },
+  },
+  resolve: {
```
Comment (jsjoeio) [inline site/vite.config.ts:24]: woah does this make it so we can do shorter imports or something?

**Review by jsjoeio — COMMENTED** [review]
Code looks great 💯 Mostly questions to understand the _why_ behind some of the changes. Gonna checkout locally and test.

**Comment by jsjoeio** [comment]
Probably user error, but how do you log in when you use a different host? I tried with GitHub but get this error: 

image

I did get it to work by manually editing `./scripts/develop.sh` to use `yarn dev:vite` though.

The first load of pages seems a bit slower (maybe the lazy load?) but other than that seems to work great! HMR is **so** fast with Vite!


https://user-images.githubusercontent.com/3806031/192026144-e625abc4-7871-4c25-909c-28ddb7799cd6.mov

**Review by jsjoeio — APPROVED** [review]
One request: update `./scripts/develop.sh` to use Vite!

File: site/index.html (hunk 1 above)
Comment (BrunoQuaresma) [reply site/index.html:1]: Vite requires the HTML file to be in the root.

File: site/src/AppRouter.tsx (hunk 3 above)
Comment (BrunoQuaresma) [reply site/src/AppRouter.tsx:16]: Yes, so the pages are not loaded in the main bundle but I'm going to change this and don't lazy load the main pages - the ones that are in the nav.

File: site/tsconfig.json (hunk 5 above)
Comment (BrunoQuaresma) [reply site/tsconfig.json:16]: Vite only works with es2020+ bundles. Webpack does not break because of that tho.

File: site/vite.config.ts (hunk 6 above)
Comment (BrunoQuaresma) [reply site/vite.config.ts:24]: Yes, unfortunately, looks like Vite is not smart enough to apply the TS aliases automatically.

**Pull request merged**